src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,510 @@
+﻿/* Copyright 2020–present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Security;
+using System.Threading.Tasks;
+using MongoDB.Bson;
+using MongoDB.Bson.Serialization;
+using MongoDB.Driver.Core.Connections;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.Core.Authentication
+{
+    /// <summary>
+    /// The Mongo AWS authenticator.
+    /// </summary>
+    public class MongoAWSAuthenticator : SaslAuthenticator
+    {
+        // constants
+        private const string AccessKeyIdExceptionMessage = "A MONGODB-AWS credential must have access key ID.";
+        private const int ClientNonceLength = 32;
+        private const string SecretAccessKeyExceptionMessage = "A MONGODB-AWS credential must have secret access key.";
```

Comment:
by @rstam 
Regarding the error message, make sure every message is a grammatically correct sentence that clearly identifies the object that is involved. So "MONGODB-AWS" doesn't seem correct (i.e. what is a "MONGODB-AWS").

Seems like the correct object is either a AwsCredential or a MongoAwsAuthenticator depending on where the error check is.

Reply:
I have changed the exception type to be `InvalidDataException`.
Regarding
> So "MONGODB-AWS" doesn't seem correct (i.e. what is a "MONGODB-AWS").

I was implementing this according to our existing code: [link](https://github.com/mongodb/mongo-csharp-driver/blob/719731cb7548ae6c1cfc253c33384084f84fb3b1/src/MongoDB.Driver/MongoCredential.cs#L532)
> `throw new ArgumentException("A MONGODB-X509 does not support a password.");`